Sally7/Internal/WireFormatting.cs

```diff
@@ -0,0 +1,67 @@
+﻿using Sally7.Protocol.S7.Messages;
+using Sally7.RequestExecutor;
+
+namespace Sally7.Internal;
+
+internal static class WireFormatting
+{
+    private const int Tpkt = 0x03_00_00_00;
+
+    /// <summary>
+    /// Data struct.
+    /// </summary>
+    /// <remarks>
+    /// This struct is only 3 bytes long, final byte is overwritten.
+    /// </remarks>
+    private const int Data = 2 << 24 // Length
+        | 0b1111_0000 << 16 // Identifier
+        | 0b1000_0000 << 8; // PDU number and EOT
+
+    private const int JobRequestHeader1 = 0x32 << 24 // Protocol ID
+        | (byte)MessageType.JobRequest << 16; // Message type
+
+    /// <summary>
+    /// The PDU reference.
+    /// </summary>
+    /// <remarks>
+    /// This value is here for legacy reasons. The <see cref="ConcurrentRequestExecutor"/> modifies the lower
+    /// byte to identify responses, the upper byte isn't actually used at the moment.
+    /// </remarks>
+    private const short PduRef = 1 << 8;
+
+    public static int WriteTpkt(ref byte destination, int length)
+    {
+        return WriteInt32(ref destination, Tpkt | length);
+    }
+
+    public static int WriteData(ref byte destination)
+    {
+        // For performance we write an int instead of separate values, the final byte of the int will be overwritten.
+        WriteInt32(ref destination, Data);
+
+        return 3;
+    }
+
+    public static int WriteJobRequestHeader(ref byte destination, int paramLength, int dataLength)
```

Comment:
```suggestion
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static int WriteJobRequestHeader(ref byte destination, int paramLength, int dataLength)
```
My gut tells me that it's worth it here.
For the JIT this method seems probably too big to inline, but the actual code is quite little, so we should give the JIT a hint.

Reply:
Do you know an easy way to verify if it will be inlined? I could at least benchmark the method call with and without `AggressiveInlining` to see if it makes a difference.

Reply:
In general I do either
* use https://sharplab.io/ and call the method from a dummy method to see what machine code gets emitted
* create a simple console app with a dummy method and set the proper env-variables (see https://github.com/dotnet/runtime/pull/74392 for more info)
* create a benchmark and use the `[DisassemblyDiagnoser]` to the see the machine code

Here with [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5il3YBLAHbYoATwDcNek1YAlDsIyD8MFgGEI+AA6CANnwDKfAG6CwMXNOoyAzE1IMAYhAg0A3jQY+mDkRgMbGIAFLBcDMDiGDAMACZWGGgMAQw6Ytj4ADIwwgDmGAAWyalx2BjYOflFAJTevl7Uvs1MAOwMAOqCsK5Q+OUq+SwdUIIxAFIQwPIwAI4ciQASMNgJUGEwEQm4SWkZ2bkFxfHllYe1Ns0AvvU+t34pykFiAPoAksL6IjBxGxFRMXiiRKT3SUEyVSOIMCZQqkNq90aLV8xHaXR60H6GEGeWGowmUxm8yWKzW70+31+4SBO2SYIh52OsLO1UKNUuvhu1C5NACfFE+jkSEcnW6MF6WJxnnuOlGpnKsUgwh2j0CABUdABrQIAXgYdAQdDsLzodBNZtNNnuAHpbQwADy4Dj4foSAB8NrtABFTgwdlAOGAMCxPdaHdanS6xOIPU1fLaw/bYK7NbhY80Eww1YVBLg/RgA0GUnmIJ9xAwHACrAx9KW8skuCJsIKq8WGBBTHwAO74mLCENxnyZ+3W5NiVPp3yywTywFKlWpH0VBh6pz2+0MUgoBh2+GFe7NQj64DkU/kc2mh0b8jCu1vBLKQSNvgH3xHugn00Wr9XhgADkkHcwwABS9ABVBhhGdYA+AYbBhDiBgAFEAHk1RoGU5QVBh50CVJJmmOYFh2ZZVj4cgV31BA7DXDctyAhhgKgCAMAgSBBTeL1Xx8I8QirGoAFkrFwbA8hgNVxB0dQCKJYjAnXBgb0Au0hNwESxIYDBJJgDDByAxNI1dGNQyzQpYlAiDwj4XJLAHDM7RHQzo0nIcHLHKAJxM7NcwYeV9AWNszNgBguGgGsYDybAwHLWA8FLARTNiR0YEVcIdQAIi0YQwA4KBYGUWTEiQhAYBy1ioHS603QYfAIDiJ9BGrIpYlrLsXz0zNW1YlIHxULgYqsHR4qsZJmoYDgdCkqBImiWJc2EAByQIoowDhm30csOFwH44MCMbarUZQ7PjBzRxgFM00wmdsNwv1CmgQJgLiDgZgiPUKIUgDdOaYgHGYYVUhGMZxK1DA/hmwFtkGcpBFLaFwtZOo9KRZE2k6fEYA+DAaPBqGmxUOGs1Bhgj0MRGOR8Hk9N+oVVXR4Gl2wcHWzx0QCeEJHmhR5E7V6NI+FCvp4MsBg2tFjG4OEOmRB2Ul2wibb6UBPziNGsyQqbFtZvlzT1dSLsDBbWIO27XtcmO5EgZiLGcepVmYcJxn2W+1HUQrCmGCpn6/pvOmrZgGSiJJMj1mpFnEnx2HhHh+kDlZeHmT3TmGm4+nreUW3NhpaH2eSQPiRI0lyOdvT7IYHJIuitWdNL3x/axm9cYjtmo5YABxGAMBQnhtrBlAamSD6Nz/dkGIshhwlT+vlEbu3m4d/sO67nvO5CJAB4YEJcHuqAMBqWOk495pp4wWes/t9n2877uuF7kIR+SLed73xPGRL1O3fIOgPa5Zp7gAbSEkUOqbxdD6BCEA+6cRQE6H0ChHQ7MBAAEE8h5FgGpGcmMKTCBEHkGoABde4NN/p+wxvnOSpEyQfC+MIH4zNtYXyjjHfYe4E6nCToiKeGMbakCbjsSOhNyHBzWOQEuqMGIVyiuIauXDgYNyQHwnOrcl43zvv3Qev4R7KRAuBCemxZHp1Pgoue/CW6livsvW+q916P23g9feLC35Hzrtwmexjz7z0violeYMH6bzsbvGor9yYu2RJ/b+9wvYoh9gDJ4J8z7/AYZ4phd0Hq+WbAsZOPhuYtAAHKdy7NATUKEoBrBMKMZsggABefA8RyLcYogR0d0n+RgGI127RcDVJgBALgT97E/1CQ8EhgNXHY14WHJJpiF7wxVm0zhtcfD5IwIUjyJSyl8EEJUmpUA6mGMzlsZJhM5ntLCZ07pvSQgBFOZ7GgVM+RQAFLTMgWZtLFT7F0+K0pqYxL0YkwE3irFgyKD5SZkMjnNNSL0oFWSGA5J+u0akYFlTYF4CwJBcQqQeOmbnTeHBrnQt7jcrk9zlD8mbM8pwyzVnFNKcYTZ2z2o5OIb7Yg254nuP+bERhhMAmBBOQs8RyKRJov9siypeRaFYsOTiphqdmhsDGFlTsu9alvFwFkMYGBDBIQQlsqWAB+IITYJBMVUGMTBAgZgqu2rq+q8FaFqRCCchgIAWmZMGbYH5zz2VjIORDblELZkZPmcjVOwrUXqDFQKQQkq6EmKUXDeVvhFUYGVXwGIuz1WauxDqvV8EGBGsVaIU1ox8AWs7FamANqYB2v1Y63AzqQ01Fde6tpnrSWZqebkZ0DBVLqXEtpVtVZvnNCESqPUBo6DkDQPcJBYBNSUSnaQWdel52akZkuw0dhV3NDAttKAm7J2GlaHcoAA==) (some copy & pasting needed) one can see the forcing the inline will help.

W/o the inline it's a tail-call (note the `jmp` as last instruction of method `Foo.Bar`, whilst w/ inline it's nice code in method `Foo.Bar_Inlined`.